src/function/scalar/list/list_lambdas.cpp

```diff
@@ -0,0 +1,417 @@
+
+		// determine if this is the lambda parameter
+		auto &bound_col_ref = (BoundColumnRefExpression &)*original;
+		if (bound_col_ref.binding.table_index == DConstants::INVALID_INDEX) {
+			is_lambda_parameter = true;
+		}
+	}
+
+	if (is_lambda_parameter) {
+		// this is a lambda parameter, so the replacement refers to the first argument, which is the list
+		replacement = make_unique<BoundReferenceExpression>(arguments[0]->alias, list_child_type, 0);
+
+	} else {
+		// this is not a lambda parameter, so we need to create a new argument for the arguments vector
+		replacement = make_unique<BoundReferenceExpression>(original->alias, original->return_type, arguments.size());
+		bound_function.arguments.push_back(original->return_type);
+		arguments.push_back(move(original));
+	}
+}
+
+static void IterateChildren(ScalarFunction &bound_function, vector<unique_ptr<Expression>> &arguments,
+                            LogicalType &list_child_type, unique_ptr<Expression> &expr) {
+
+	if (expr->expression_class == ExpressionClass::BOUND_SUBQUERY) {
+		throw InvalidInputException("Subqueries are not supported in lambda expressions!");
+	}
+
+	// these expression classes do not have children, transform them
+	if (expr->expression_class == ExpressionClass::BOUND_CONSTANT ||
+	    expr->expression_class == ExpressionClass::BOUND_COLUMN_REF ||
+	    expr->expression_class == ExpressionClass::BOUND_DEFAULT ||
+	    expr->expression_class == ExpressionClass::BOUND_PARAMETER ||
+	    expr->expression_class == ExpressionClass::BOUND_REF) {
+
+		// move the expr because we are going to replace it
+		auto original = move(expr);
+		unique_ptr<Expression> replacement;
+
+		TransformExpression(original, replacement, bound_function, arguments, list_child_type);
+
+		// replace the expression
+		expr = move(replacement);
+
+	} else {
+		// recursively enumerate the children of the expression
+		ExpressionIterator::EnumerateChildren(*expr, [&](unique_ptr<Expression> &child) {
+			IterateChildren(bound_function, arguments, list_child_type, child);
+		});
+	}
+}
+
+static unique_ptr<FunctionData> ListLambdaBind(ClientContext &context, ScalarFunction &bound_function,
+                                               vector<unique_ptr<Expression>> &arguments) {
+
+	// remove the lambda function
+	auto lambda_expr = move(arguments.back());
+	arguments.pop_back();
+	bound_function.arguments.pop_back();
+
+	if (arguments[0]->return_type.id() == LogicalTypeId::SQLNULL) {
+		bound_function.arguments[0] = LogicalType::SQLNULL;
+		bound_function.return_type = LogicalType::SQLNULL;
+		return make_unique<VariableReturnBindData>(bound_function.return_type);
+	}
+
+	D_ASSERT(arguments[0]->return_type.id() == LogicalTypeId::LIST);
+	auto list_child_type = ListType::GetChildType(arguments[0]->return_type);
+
+	// iterate and transform the children of the lambda expression
+	IterateChildren(bound_function, arguments, list_child_type, lambda_expr);
```

Comment:
Can this be made more generic and moved to the `Binder`?

Also the name of this function (`IterateChildren`) could be a bit more descriptive :)